README.md

# Big Red Button™

A big red button with an Arduino Leonardo (MEGA32U4) acting as a keyboard.

This code is an extended version of the Keyboard library already present in the Arduino IDE. What the official version lacks is the support for **Page 0x01** keys (like Volume Up/Down, Mute, Sleep, Power Off), and the support for reading LED status.

The intention with this library was to build a big red button that acts like a keyboard, allowing a software in the background to listen on the pressed keys and do things when that happen. I mainly use it as a party tool to make sound effects, like for a gaming event. I wrote an app for this specific purpose, and it's also publicly available: [https://vbstudio.hu/apps/vasparittya](https://vbstudio.hu/apps/vasparittya).

You can also use it to...
- Trigger stuff that otherwise require multiple keys pressed simultaneously, like taking screenshots and recording video with ***Xbox Game Bar*** or ***GeForce Experience*** (see "[More use-case examples](#more-use-case-examples)" below).
- Host quiz shows, given you have more than one button. This would require a listener app that can tell who pressed their button first.
- Program macros.

To read about the hardware part of the project, visit my [article](https://vbstudio.hu/blog/20200905-A-Big-Red-Button-Acting-as-a-Keyboard-Using-Arduino).

## How to install code
- Download and install the [Arduino IDE](https://www.arduino.cc/en/software).
- The Arduino IDE by default stores all Sketches in a folder named **"Arduino"** in your **Documents** folder.
- Download the repository and copy everything from the **"Source Code"** folder to there.
	- It might ask whether you want to merge the **"libraries"** folder with the new one, choose yes.
- Start the Arduino IDE and select **"File"**, **"Sketches"**, and then choose **"BigRedButton"** from the list.

This sketch file contains all the relevant part to change and tweak behavior or switch I/O pins. The "under the hood" parts are located in the C++ classes under the "libraries" folder.

## Build and upload sketch
- After the Leonardo board is plugged in, go to **"Tools"**, **"Board"**, then select **"Arduino Leonardo"**.
- Go to **"Tools"** again, then **"Port"**, and select the port that looks like **"COM# (Arduino Leonardo)"**.
- Press and **hold the reset button** on the Leonardo board. (Add one if it does not have it)
- **Start Upload** with the right-arrow icon on the menu bar.
- As soon as the console at the bottom says **"Uploading..."**, release the reset button.
- It should finish in a few seconds, and then done.

## Changing keys
The `Keyboard` class specifies separate function calls for **page 0x01** and **page 0x07** keys.
Constants for the most common key codes and modifier keys are defined in `VbsKeyboard.h`.

### Page 0x07 key calls (regular keys)
``` c++
Keyboard.PressKey(uint8_t key, uint8_t modifier = MOD_NONE)
```
Issues a key press and then immediately a release for the specified `key`. One or multiple `modifier`-s can be specified to be pressed along with the key, like Ctrl, Alt, Shift or GUI (Window key). See `VbsKeyboard.h` for the available constants.

``` c++
Keyboard.HoldKey(uint8_t key, uint8_t modifier = MOD_NONE)
```
Same as `PressKey()`, except the key and modifiers remains pressed until `ReleaseKey()` is called.

``` c++
Keyboard.ReleaseKey()
```
Releases all currently pressed (page 0x07) keys.

### Page 0x01 key calls (system and media keys)
``` c++
Keyboard.PressKeyPage1(uint16_t key)
```
Issues a key press and then immediately a release for the specified page 0x01 `key`. No holding, or modifiers for this one.

## More use-case examples
Overwrite any of the preset programs with these.

### GeForce Experience screenshot/recording
Short press to take screenshot, long press to start/stop video recording.
``` c++
auto event = BigRedButton.PollDualButtonEvent();
if (event.Click) Keyboard.PressKey(KEY_F1, MOD_LEFT_ALT);
if (event.LongPress) Keyboard.PressKey(KEY_F9, MOD_LEFT_ALT);
```

### Xbox Game Bar screenshot/recording
Short press to take screenshot, long press to start/stop video recording.
``` c++
auto event = BigRedButton.PollDualButtonEvent();
if (event.Click) Keyboard.PressKey(KEY_PRINT_SCREEN, MOD_LEFT_GUI | MOD_LEFT_ALT);
if (event.LongPress) Keyboard.PressKey(KEY_R, MOD_LEFT_GUI | MOD_LEFT_ALT);
```

## License
MIT

changelog.txt
v2.0
- Moved all "under the hood" parts into a separate class to make top-level code simpler.
- Added LED pulse size option.
- Completely separated simple, dual and quad button event handling and polling to allow fine control over LED behavior.
- Added flashing to indicate long press event.
- Also added flashing to clicks when button is constantly lit.

v1.2
- Added LED brightness constant.
- Added pulsing effect to LED when activated with Scroll Lock.
- Increased double click duration to 400ms.
- Decreased long press time to 700ms.

v1.1
- Fixed Schmitt trigger to be symmetrical.
- Removed Schmitt trigger from swithes.

v1.0
- Initial version.

Source Code/libraries/VbsKeyboard/VbsKeyboard.cpp
/*
    VbsKeyboard.cpp
    
    Copyright (c) 2020, Balazs Vecsey, www.vbstudio.hu
    
    This is a merged, modified and extended version of the Keyboard.cpp
    and HID.cpp from the official Arduino library, version 1.8.12.
    Original copyright is as follows:
    
    Copyright (c) 2015, Arduino LLC
    Original code (pre-library): Copyright (c) 2011, Peter Barrett
    
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.
    
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
*/

#include "VbsKeyboard.h"

#if defined(USBCON)

#define HID_REPORTID_KEYBOARD       0x02
#define HID_REPORTID_GENERICDESKTOP 0x04

static const uint8_t _hidReportDescriptorPage1[] PROGMEM = {
    0x05, 0x01,                                 // USAGE_PAGE (Generic Desktop)
    0x09, 0x06,                                 // USAGE (Keyboard)
    0xA1, 0x01,                                 // COLLECTION (application)
    0x85, HID_REPORTID_GENERICDESKTOP,          // REPORT_ID (HID_REPORTID_CONSUMERCONTROL)
    
    // 4 Media Keys
    0x15, 0x00,                                 // LOGICAL_MINIMUM
    0x26, 0xFF, 0x03,                           // LOGICAL_MAXIMUM (3ff)
    0x19, 0x00,                                 // USAGE_MINIMUM (0)
    0x2A, 0xFF, 0x03,                           // USAGE_MAXIMUM (3ff)
    0x95, 0x04,                                 // REPORT_COUNT (4)
    0x75, 0x10,                                 // REPORT_SIZE (16)
    0x81, 0x00,                                 // INPUT (Data,Ary,Abs)
    0xC0 // END_COLLECTION
};

static const uint8_t _hidReportDescriptorPage7[] PROGMEM = {
    0x05, 0x01, // USAGE_PAGE (Generic Desktop)
    0x09, 0x06, // USAGE (Keyboard)
    0xa1, 0x01, // COLLECTION (Application)
    0x85, HID_REPORTID_KEYBOARD, // REPORT_ID (2)
    
    0x05, 0x07, //   USAGE_PAGE (Keyboard)
    
    // Keyboard Modifiers (shift, alt, ...)
    0x19, 0xe0, //   USAGE_MINIMUM (Keyboard LeftControl)
    0x29, 0xe7, //   USAGE_MAXIMUM (Keyboard Right GUI)
    0x15, 0x00, //   LOGICAL_MINIMUM (0)
    0x25, 0x01, //   LOGICAL_MAXIMUM (1)
    0x75, 0x01, //   REPORT_SIZE (1)
    0x95, 0x08, //   REPORT_COUNT (8)
    0x81, 0x02, //   INPUT (Data,Var,Abs)
    
    0x95, 0x01, //   REPORT_COUNT (1)
    0x75, 0x08, //   REPORT_SIZE (8)
    
    0x81, 0x03, //   INPUT (Cnst,Var,Abs)
    0x95, 0x05, //   REPORT_COUNT (5)
    0x75, 0x01, //   REPORT_SIZE (1)
    
    // 3 LEDs
    0x05, 0x08, //   USAGE_PAGE (LEDs)
    0x19, 0x01, //   USAGE_MINIMUM (1)
    0x29, 0x05, //   USAGE_MAXIMUM (5)
    0x91, 0x02, //   OUTPUT (Data,Var,Abs) // LED report
    0x95, 0x01, //   REPORT_COUNT (1)
    0x75, 0x03, //   REPORT_SIZE (3)
    
    0x91, 0x01, //   OUTPUT (Constant) // padding 
    0x95, 0x06, //   REPORT_COUNT (6)
    // END of LEDs
    
    // 6 Keyboard keys
    0x95, 0x06, //   REPORT_COUNT (6)
    0x75, 0x08, //   REPORT_SIZE (8)
    0x15, 0x00, //   LOGICAL_MINIMUM (0)
    0x25, 0x73, //   LOGICAL_MAXIMUM (115)
    0x05, 0x07, //   USAGE_PAGE (Keyboard)
    0x19, 0x00, //   USAGE_MINIMUM (Reserved (no event indicated))
    0x29, 0x73, //   USAGE_MAXIMUM (Keyboard Application)
    0x81, 0x00, //   INPUT (Data,Ary,Abs)
    0xc0,       // END_COLLECTION
};

// Implementation of PluggableUSBModule
int VbsKeyboard::getInterface(uint8_t* interfaceCount)
{
    *interfaceCount += 1; // uses 1
    HIDDescriptor hidInterface = {
        D_INTERFACE(pluggedInterface, 1, USB_DEVICE_CLASS_HUMAN_INTERFACE, HID_SUBCLASS_NONE, HID_PROTOCOL_NONE),
        D_HIDREPORT(_descriptorSize),
        D_ENDPOINT(USB_ENDPOINT_IN(pluggedEndpoint), USB_ENDPOINT_TYPE_INTERRUPT, USB_EP_SIZE, 0x01)
    };
    return USB_SendControl(0, &hidInterface, sizeof(hidInterface));
}

// Implementation of PluggableUSBModule
int VbsKeyboard::getDescriptor(USBSetup& setup)
{
    // Check if this is a HID Class Descriptor request
    if (setup.bmRequestType != REQUEST_DEVICETOHOST_STANDARD_INTERFACE) return 0;
    if (setup.wValueH != HID_REPORT_DESCRIPTOR_TYPE) return 0;
    
    // In a HID Class Descriptor wIndex cointains the interface number
    if (setup.wIndex != pluggedInterface) return 0;
    
    int total = 0;
    HIDSubDescriptor* node;
    for (node = _rootNode; node; node = node->next)
    {
        int res = USB_SendControl(TRANSFER_PGM, node->data, node->length);
        if (res == -1) return -1;
        total += res;
    }
    
    // Reset the protocol on reenumeration. Normally the host should not assume the state of the protocol
    // due to the USB specs, but Windows and Linux just assumes its in report mode.
    _protocol = HID_REPORT_PROTOCOL;
    
    return total;
}

// Implementation of PluggableUSBModule
uint8_t VbsKeyboard::getShortName(char *name)
{
    name[0] = 'H';
    name[1] = 'I';
    name[2] = 'D';
    name[3] = 'A' + (_descriptorSize & 0x0F);
    name[4] = 'A' + ((_descriptorSize >> 4) & 0x0F);
    return 5;
}

// Implementation of PluggableUSBModule
bool VbsKeyboard::setup(USBSetup& setup)
{
    if (pluggedInterface != setup.wIndex) return false;
    
    uint8_t request = setup.bRequest;
    uint8_t requestType = setup.bmRequestType;
    
    if (requestType == REQUEST_DEVICETOHOST_CLASS_INTERFACE)
    {
        if (request == HID_GET_REPORT)
        {
            // TODO: HID_GetReport();
            return true;
        }
        else if (request == HID_GET_PROTOCOL)
        {
            // TODO: Send8(_protocol);
            return true;
        }
        else if (request == HID_GET_IDLE)
        {
            // TODO: Send8(_idle);
        }
    }
    
    if (requestType == REQUEST_HOSTTODEVICE_CLASS_INTERFACE)
    {
        if (request == HID_SET_PROTOCOL)
        {
            // The USB Host tells us if we are in boot or report mode.
            // This only works with a real boot compatible device.
            _protocol = setup.wValueL;
            return true;
        }
        else if (request == HID_SET_IDLE)
        {
            _idle = setup.wValueL;
            return true;
        }
        else if (request == HID_SET_REPORT)
        {
            if (setup.wLength == 2) 
            {
                uint8_t data[2];
                if (USB_RecvControl(data, 2) == 2) 
                {
                    _ledsState = data[1];
                }
            }
        }
    }
    
    return false;
}

VbsKeyboard::VbsKeyboard(void) :
    PluggableUSBModule(1, 1, _epType),
    _rootNode(NULL), _descriptorSize(0),
    _protocol(HID_REPORT_PROTOCOL), _idle(1)
{
    _epType[0] = EP_TYPE_INTERRUPT_IN;
    PluggableUSB().plug(this);
    
    // Append generic keyboard descriptor
    static HIDSubDescriptor nodeKeyboard(_hidReportDescriptorPage7, sizeof(_hidReportDescriptorPage7));
    AppendDescriptor(&nodeKeyboard);
    
    // Append system keyboard descriptor
    static HIDSubDescriptor nodeConsumer(_hidReportDescriptorPage1, sizeof(_hidReportDescriptorPage1));
    AppendDescriptor(&nodeConsumer);
}

void VbsKeyboard::AppendDescriptor(HIDSubDescriptor* node)
{
    if (!_rootNode)
    {
        _rootNode = node;
    }
    else
    {
        HIDSubDescriptor *current = _rootNode;
        while (current->next)
        {
            current = current->next;
        }
        current->next = node;
    }
    _descriptorSize += node->length;
}

#define SendReportPage1() { SendReport(HID_REPORTID_GENERICDESKTOP, &_keyReportPage1, sizeof(KeyReportPage1)); }
#define SendReportPage7() { SendReport(HID_REPORTID_KEYBOARD, &_keyReportPage7, sizeof(KeyReportPage7)); }

void VbsKeyboard::SendReport(uint8_t id, void* data, int len)
{
    if (USB_Send(pluggedEndpoint, &id, 1) < 0) return;
    if (USB_Send(pluggedEndpoint | TRANSFER_RELEASE, data, len) < 0) return;
}

void VbsKeyboard::PressKeyPage1(uint16_t key) 
{
    // Press
    _keyReportPage1.keys[0] = key;
    _keyReportPage1.keys[1] = 0;
    _keyReportPage1.keys[2] = 0;
    _keyReportPage1.keys[3] = 0;
    SendReportPage1();
    
    // Release
    _keyReportPage1.keys[0] = 0;
    SendReportPage1();
}
    
void VbsKeyboard::PressKey(uint8_t key, uint8_t modifier)
{
    HoldKey(key, modifier);
    ReleaseKey();
}

void VbsKeyboard::HoldKey(uint8_t key, uint8_t modifier)
{
    _keyReportPage7.keys[0] = key;
    _keyReportPage7.keys[1] = 0;	
    _keyReportPage7.keys[2] = 0;
    _keyReportPage7.keys[3] = 0;
    _keyReportPage7.keys[4] = 0;
    _keyReportPage7.keys[5] = 0;	
    _keyReportPage7.modifiers = modifier;
    SendReportPage7();
}

bool VbsKeyboard::GetLedState(uint8_t mask) const
{
    return _ledsState & mask;
}


VbsKeyboard Keyboard;

#endif /* if defined(USBCON) */

Source Code/libraries/VbsBigRedButton/VbsBigRedButton.h
/*
    MIT License
    
    Copyright (c) 2021, Balazs Vecsey, www.vbstudio.hu
    
    Permission is hereby granted, free of charge, to any person obtaining a copy of
    this software and associated documentation files (the "Software"), to deal in the
    Software without restriction, including without limitation the rights to use,
    copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
    Software, and to permit persons to whom the Software is furnished to do so,
    subject to the following conditions:
    
    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.
       
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
    FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
    COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
    AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
    WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#ifndef VBS_BIG_RED_BUTTON_h
#define VBS_BIG_RED_BUTTON_h

#include <Arduino.h>
#include <VbsKeyboard.h>

struct VbsSingleButtonEvent
{
    bool Press;
    bool Release;
};

struct VbsDualButtonEvent
{
    bool Click;
    bool LongPress;
};

struct VbsQuadButtonEvent
{
    bool SingleClick;
    bool DoubleClick;
    bool LongPress;
    bool LongPressDoubleClick;
};

class VbsBigRedButton
{
private:
    enum LightOverride
    {
        LIGHT_FREE = 0,
        LIGHT_ON = 1,
        LIGHT_OFF = 2
    };
    
    // PINS
    const uint8_t _pinButton;
    const uint8_t _pinLight;
    const uint8_t _pinSwitch1;
    const uint8_t _pinSwitch2;
    
    // CONFIG
    int _longPressTime = 700;
    int _doubleClickTime = 400;
    float _lightChangeSpeed = 25.0f; // (bigger value -> faster transition)
    int _lightFeedbackFlashSpeed = 150;
    float _lightMaxBrightness = 1.0f;
    bool _lightPulseEnabled = true;
    float _lightPulseFreq = 0.5f; // Hz
    float _lightPulseSize = 0.1; // %
    
    // STATE
    int _programIndex;
    
    bool _buttonLastState;
    unsigned long _longPressStarted;
    unsigned long _doubleClickStarted;
    bool _doubleClickInProgress;
    bool _nextReleaseIsDoubleClick;
    bool _longPressFired;
    
    bool _lightKeepLit = false;
    bool _lightFeedbackFlashRunning = false;
    unsigned long _lightFeedbackFlashTime = 0;
    LightOverride _lightOverride = LIGHT_FREE;
    float _lightBrightness = 0.0f;
    float _lightPulseTime = 0.0f;
    unsigned long _lastTimestamp = 0;
    
    // FUNCTIONS
    int readProgramSwitch() const;
    bool readButton() const;
    
    void resetButtonState();
    void updateLight();
    void triggerFeedbackFlash();
    
public:
    VbsBigRedButton(const uint8_t pinButton, const uint8_t pinLight, const uint8_t pinSwitch1, const uint8_t pinSwitch2);
    
    void SetLongPressTime(const int ms);
    void SetDoubleClickTime(const int ms);
    void SetLightChangeSpeed(const float speed);
    void SetLightFeedbackFlashSpeed(const int ms);
    void SetLightMaxBrightness(const float brightness);
    void SetLightPulse(const float frequency, const float size = 0.1f);
    
    void KeepLightLit(const bool lit);
    
    int GetProgramIndex();
    VbsSingleButtonEvent PollSingleButtonEvent();
    VbsDualButtonEvent PollDualButtonEvent();
    VbsQuadButtonEvent PollQuadButtonEvent();
};

#endif

Source Code/libraries/VbsKeyboard/VbsKeyboard.h
/*
    VbsKeyboard.h
    
    Copyright (c) 2020, Balazs Vecsey, www.vbstudio.hu
    
    This is a merged, modified and extended version of the Keyboard.h
    and HID.h from the official Arduino library, version 1.8.12.
    Original copyright is as follows:
    
    Copyright (c) 2015, Arduino LLC
    Original code (pre-library): Copyright (c) 2011, Peter Barrett
    
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.
    
    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.
*/

#ifndef VBS_KEYBOARD_h
#define VBS_KEYBOARD_h


#include <stdint.h>
#include <Arduino.h>
#include "PluggableUSB.h"


#if defined(USBCON)


// HID 'Driver'
// ----------------------------------------------
#define HID_GET_REPORT        0x01
#define HID_GET_IDLE          0x02
#define HID_GET_PROTOCOL      0x03
#define HID_SET_REPORT        0x09
#define HID_SET_IDLE          0x0A
#define HID_SET_PROTOCOL      0x0B

#define HID_HID_DESCRIPTOR_TYPE         0x21
#define HID_REPORT_DESCRIPTOR_TYPE      0x22
#define HID_PHYSICAL_DESCRIPTOR_TYPE    0x23

// HID subclass HID1.11 Page 8 4.2 Subclass
#define HID_SUBCLASS_NONE 0
#define HID_SUBCLASS_BOOT_INTERFACE 1

// HID Keyboard/Mouse bios compatible protocols HID1.11 Page 9 4.3 Protocols
#define HID_PROTOCOL_NONE 0
#define HID_PROTOCOL_KEYBOARD 1
#define HID_PROTOCOL_MOUSE 2

// Normal or bios protocol (Keyboard/Mouse) HID1.11 Page 54 7.2.5 Get_Protocol Request
// "protocol" variable is used for this purpose.
#define HID_BOOT_PROTOCOL	0
#define HID_REPORT_PROTOCOL	1

// HID Request Type HID1.11 Page 51 7.2.1 Get_Report Request
#define HID_REPORT_TYPE_INPUT   1
#define HID_REPORT_TYPE_OUTPUT  2
#define HID_REPORT_TYPE_FEATURE 3

typedef struct
{
    uint8_t len;      // 9
    uint8_t dtype;    // 0x21
    uint8_t addr;
    uint8_t versionL; // 0x101
    uint8_t versionH; // 0x101
    uint8_t country;
    uint8_t desctype; // 0x22 report
    uint8_t descLenL;
    uint8_t descLenH;
} HIDDescDescriptor;

typedef struct 
{
    InterfaceDescriptor hid;
    HIDDescDescriptor desc;
    EndpointDescriptor in;
} HIDDescriptor;

class HIDSubDescriptor {
public:
    HIDSubDescriptor *next = NULL;
    HIDSubDescriptor(const void *d, const uint16_t l) : data(d), length(l) { }

    const void* data;
    const uint16_t length;
};

#define D_HIDREPORT(length) { 9, 0x21, 0x01, 0x01, 0, 1, 0x22, lowByte(length), highByte(length) }


// Keyboard
// ----------------------------------------------
// LEDs
#define KB_LED_NUM_LOCK     0x01
#define KB_LED_CAPS_LOCK    0x02
#define KB_LED_SCROLL_LOCK  0x04

// Modifiers (page 7)
#define MOD_NONE         0x00
#define MOD_LEFT_CTRL    0x01
#define MOD_LEFT_SHIFT   0x02
#define MOD_LEFT_ALT     0x04
#define MOD_LEFT_GUI     0x08
#define MOD_RIGHT_CTRL   0x10
#define MOD_RIGHT_SHIFT  0x20
#define MOD_RIGHT_ALT    0x40
#define MOD_RIGHT_GUI    0x80

// Modifier keys (page 7)
#define KEY_LEFT_CTRL    0x80
#define KEY_LEFT_SHIFT   0x81
#define KEY_LEFT_ALT     0x82
#define KEY_LEFT_GUI     0x83
#define KEY_RIGHT_CTRL   0x84
#define KEY_RIGHT_SHIFT  0x85
#define KEY_RIGHT_ALT    0x86
#define KEY_RIGHT_GUI    0x87

// Character keys (page 7) (some of these are language specific!)
#define KEY_SPACE        0x2c
#define KEY_A            0x04
#define KEY_B            0x05
#define KEY_C            0x06
#define KEY_D            0x07
#define KEY_E            0x08
#define KEY_F            0x09
#define KEY_G            0x0a
#define KEY_H            0x0b
#define KEY_I            0x0c
#define KEY_J            0x0d
#define KEY_K            0x0e
#define KEY_L            0x0f
#define KEY_M            0x10
#define KEY_N            0x11
#define KEY_O            0x12
#define KEY_P            0x13
#define KEY_Q            0x14
#define KEY_R            0x15
#define KEY_S            0x16
#define KEY_T            0x17
#define KEY_U            0x18
#define KEY_V            0x19
#define KEY_W            0x1a
#define KEY_X            0x1b
#define KEY_Y            0x1c
#define KEY_Z            0x1d

// Function keys (page 7)
#define KEY_ENTER        0x28
#define KEY_ESC          0x29
#define KEY_BACKSPACE    0x2a
#define KEY_TAB          0x2b
#define KEY_PRINT_SCREEN 0x46
#define KEY_F1           0x3A
#define KEY_F2           0x3B
#define KEY_F3           0x3C
#define KEY_F4           0x3D
#define KEY_F5           0x3E
#define KEY_F6           0x3F
#define KEY_F7           0x40
#define KEY_F8           0x41
#define KEY_F9           0x42
#define KEY_F10          0x43
#define KEY_F11          0x44
#define KEY_F12          0x45
#define KEY_F13          0x68
#define KEY_F14          0x69
#define KEY_F15          0x6a
#define KEY_F16          0x6b
#define KEY_F17          0x6c
#define KEY_F18          0x6d
#define KEY_F19          0x6e
#define KEY_F20          0x6f
#define KEY_F21          0x70
#define KEY_F22          0x71
#define KEY_F23          0x72
#define KEY_F24          0x73

// Page 1 keys
#define KEY1_SYSTEM_SLEEP 0x82


// Low level key report: up to 6 keys and shift, ctrl etc at once
typedef struct
{
    uint8_t modifiers;
    uint8_t reserved;
    uint8_t keys[6];
} KeyReportPage7;

typedef union {
    // Every usable Consumer key possible, up to 4 keys presses possible
    uint8_t whole8[0];
    uint16_t whole16[0];
    uint32_t whole32[0];
    uint16_t keys[4];
} KeyReportPage1;


class VbsKeyboard : public PluggableUSBModule
{
public:
    VbsKeyboard(void);
    
    // Page 0x01 (Generic Desktop)
    void PressKeyPage1(uint16_t key);
    
    // Page 0x07 (Keyboard/Keypad)
    void PressKey(uint8_t key, uint8_t modifier = MOD_NONE);
    void HoldKey(uint8_t key, uint8_t modifier = MOD_NONE);
    inline void ReleaseKey() { HoldKey(0, 0); }
    
    bool GetLedState(uint8_t mask) const;
    
protected:
    // Implementation of the PluggableUSBModule
    int getInterface(uint8_t* interfaceCount);
    int getDescriptor(USBSetup& setup);
    bool setup(USBSetup& setup);
    uint8_t getShortName(char* name);
    
private:
    // HID
    uint8_t _epType[1];
    HIDSubDescriptor* _rootNode;
    uint16_t _descriptorSize;
    uint8_t _protocol;
    uint8_t _idle;
    
    // Keyboard
    KeyReportPage1 _keyReportPage1;
    KeyReportPage7 _keyReportPage7;
    uint8_t _ledsState;
    
    void SendReport(uint8_t id, void* data, int len);
    
    void AppendDescriptor(HIDSubDescriptor* node);
};

// Singleton instance
extern VbsKeyboard Keyboard;

#endif // USBCON
#endif

